M3DB Terraform Recipe

dewan-ahmed opened the pull request:
# What changed, and why it matters

Resolves https://github.com/aiven/devportal/issues/872

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,115 @@
+Monitor PostgreSQL® metrics using M3-related services
+=====================================================
+
+Aiven for M3DB is a powerful metrics engine that can be used to monitor your Aiven services on a very large scale. Aiven for M3 Aggregator can store your data at various resolutions for different workloads at scale. You can monitor a relational database like PostgreSQL by sending its metrics to M3DB with M3 Aggregator integration and then visualizing the metrics data on an Aiven for Grafana® dash ...
```
Is it an engine or a database? Also, you're using it only for the monitoring of Aiven services but can be used for any time series data scenario

```suggestion
Aiven for M3DB is a powerful metrics database that can be used to monitor your Aiven services on a very large scale. Aiven for M3 Aggregator can store your data at various resolutions for different workloads at scale. You can monitor a relational database like PostgreSQL by sending its metrics to M3DB with M3 Aggregator integration and then visualizing the metrics data on an Aiven for Grafana® dashboard.
```

dewan-ahmed replied:
Hi @ftisiot good point. Using PG as source made it very similar to the other recipe. I'll change the source.

On the "engine" or "database", I took the wording from the landing page of https://m3db.io. I think metrics engine or time-series database - both are Ok. But I can swap metrics engine with time-series database if you prefer.

dewan-ahmed replied:
Similar wording is found here: https://www.youtube.com/watch?v=YKXVxbvWW3k

I believe that M3, as a whole, is a metrics engine whereas M3DB (a subset) is the time-series database.

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,115 @@
+
+.. mermaid::
+
+  graph LR
```
Also this looks equal to https://developer.aiven.io/docs/tools/terraform/reference/cookbook/grafana-m3db-postgresql-recipe.html with only the addition of M3Agg... should we cover some other techs?

ftisiot reviewed (changes requested):
Hi Dewan! Good one, but I feel this is almost the exact copy of https://developer.aiven.io/docs/tools/terraform/reference/cookbook/grafana-m3db-postgresql-recipe.html can we use a different source?

dewan-ahmed commented:
Hi @ftisiot, 

https://github.com/aiven/devportal/pull/1046/commits/e3da9ce9bda64f9fb93fa1ffa1267e6b28107793 brings in the following fixes

- Removal of PostgreSQL and Grafana from text and diagram
- Adding link to M3 and M3 Aggregator in the intro
- Fixing "metrics engine" --> "time-series database"
- Terraform - Removal of PG, Grafana and service integrations. Added aggregated namespaces with 3 resolutions/retentions
- Added explanation for the `namespaces` config

https://github.com/aiven/devportal/pull/1046/commits/de6c0a34aecab129e2e7014a90a0537da9c985be and https://github.com/aiven/devportal/pull/1046/commits/66bfb133857f8deba5b4b2c1536269de4fb0f36d fix a broken link and removes extra '_' from the links. Although this file is not related to the PR, it was making the CI run fail
`(docs/tools/terraform/howto/upgrade-provider-v1-v2: line   93) broken    https://www.terraform.io/cli/config/config-file#provider-installation - Anchor 'provider-installation' not found`

ftisiot commented on _toc.yml:
```diff
@@ -138,9 +138,9 @@ entries:
               title: PostgreSQL with custom configs
             - file: docs/tools/terraform/reference/cookbook/kafka-mongodb-recipe
-              title: Apache Kafka MongoDB Source Connector  
-            - file: docs/tools/terraform/reference/cookbook/kafka-custom-conf-recipe.rst
```
Why are we removing the custom configuration one?

dewan-ahmed replied:
That recipe and TOC entry weren't there when this PR went in. Re-added back. Thanks.

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,105 @@
+      M3Agg[M3 Aggregator]
+      M3 -.-> M3Agg
+
+In the above diagram, the M3 service contains both M3DB and M3 Coordinator. If you're planning for longer retentions, the M3 aggregator can help reduce the volume of time series data stored by using variable data point resolutions. 
```
This is technically not correct.
You can do aggregation even without the M3aggregator node. But if you plan to do a lot, having an m3 aggregator node will help

ftisiot reviewed (changes requested):
Almost there! 
Some comments on the wording, we might need to change the examples and give a bit more info when we state something (like aggregation data retention)

dewan-ahmed commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,105 @@
+Setup M3-related services using Aiven Terraform Provider
+========================================================
+
+`Aiven for M3DB <https://aiven.io/m3>`_ is a powerful time-series database that can be used to monitor your Aiven services on a very large scale. `Aiven for M3 Aggregator <https://aiven.io/m3-aggregator>`_ can store your data at various resolutions for different workloads at scale. You can monitor a relational database like PostgreSQL® by sending its metrics to M3DB with M3 Aggregator integration ...
```
I removed the wordings around monitoring and focused on time-series and IoT (as example).

dewan-ahmed commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,105 @@
+  }
+
+``namespaces`` in M3 is used to determine how metrics are stored and retained. There is always one unaggregated namespace which is configured under the ``demo-m3db`` resource ``namespaces`` block. There are three aggregated namespaces defined within the same block for different resolution settings.
+Typically, a low resolution data is retained for a longer period of time than a high resolution data. 
```
Added another line to explain why high res data takes more storage.

dewan-ahmed commented:
Hi @ftisiot thanks for these insightful review comments. https://github.com/aiven/devportal/pull/1046/commits/973b4d1416d73847744a20e4502b218c5995fdd1 fixes:

- the missing _toc.yml entry
- First paragraph texts s/monitor/time-series(iot)
- explanation that aggregation can be done without M3 aggregator and when it makes sense to have dedicated M3 aggregator
- explanation why high resolution data is paired with low retention value

Thanks and looking forward to fixing any issues that might still persist.

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,108 @@
+Setup M3-related services using Aiven Terraform Provider
+========================================================
+
+`Aiven for M3DB <https://aiven.io/m3>`_ is a powerful time-series database that can be used when handling very large volumes of metrics and scalability is a concern. `Aiven for M3 Aggregator <https://aiven.io/m3-aggregator>`_ can store your data at various resolutions for different workloads at scale. 
```
Should we link to devportal rather than product pages?

dewan-ahmed replied:
I was contemplating on that too. Since we're already on the DevPortal and we're talking about products here (not the open-source project), my vote would be to link the product page. But I can change if you feel strongly about this.

ftisiot replied:
I believe what we did as of now is to keep people informed. If we have a devportal article I would move the link to it

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,108 @@
+      M3 -.-> M3Agg
+
+In the above diagram, the M3 service contains both M3DB and M3 Coordinator. The service integration between M3DB + M3 Coordinator and M3 Aggregator brings unaggregated metrics from M3 Coordinator to M3 Aggregator. 
+While you can do aggregation without an M3 Aggregator node, a dedicated metrics aggregator can help reduce the volume of time series data stored by using variable data point resolutions. This is especially true if you want the ability to automatically roll-up data over time.
```
```suggestion
While you can perform aggregations without an M3 Aggregator node, a dedicated metrics aggregator can help in cases when the downsampling workload is slowing down the ingestion of metrics by the main M3DB.
```

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,108 @@
+    destination_service_name = aiven_m3aggregator.demo-m3a.service_name
+  }
+
+``namespaces`` in M3 is used to determine how metrics are stored and retained. There is always one unaggregated namespace which is configured under the ``demo-m3db`` resource ``namespaces`` block. There are three aggregated namespaces defined within the same block for different resolution settings.
```
When you mention the 3 aggregated namespaces it would be good to reference the names and what each one does like
* `m3_highres_ns` downsamples the data to 10s and keeps the data (retention time) for 1h
etc

dewan-ahmed replied:
Good call. Sending that improvement via next commit.

ftisiot reviewed (changes requested):
Smaller changes, I'm currently testing the recipe. Will let you know itf it works!

dewan-ahmed commented:
Hi @ftisiot I had to tweak the resolution/retention values because business-8 plan couldn't handle the previous combination (the retention time for those resolution were too high). One could still get those combinations to work with higher plans but for sample code/recipe, I decided to lower the retention setting so that the audience can test in an affordable way :)

ftisiot commented on docs/tools/terraform/reference/cookbook/m3db-m3agg-recipe.rst:
```diff
@@ -0,0 +1,122 @@
+``namespaces`` in M3 is used to determine how metrics are stored and retained. There is always one unaggregated namespace which is configured under the ``demo-m3db`` resource ``namespaces`` block. There are three aggregated namespaces defined within the same block for different resolution settings.
+
+- ``m3_default_unaggregated_ns`` keeps the unaggregated data for 2h (retention time)
+- ``m3_lowRes_aggregated_ns`` downsamples the data to 10m (resolution) and keeps the data for 6d (retention time)
```
downsamples doesn't seem to be a world, we can change it to aggregates

ftisiot replied:
Same for the 3 downsamples below

ftisiot reviewed (changes requested):
Last couple of changes and we can merge

ftisiot reviewed (approved):
LGTM
